splitgraph/ingestion/sqlite/__init__.py

```diff
@@ -0,0 +1,177 @@
+    if "CHAR" in sqlite_type or "CLOB" in sqlite_type or "TEXT" in sqlite_type:
+        return "TEXT"
+    # If the declared type for a column contains the string "BLOB" or if no type is specified then the column has affinity BLOB.
+    if "BLOB" in sqlite_type:
+        return "BLOB"
+    # If the declared type for a column contains any of the strings "REAL", "FLOA", or "DOUB" then the column has REAL affinity.
+    if "REAL" in sqlite_type or "FLOA" in sqlite_type or "DOUB" in sqlite_type:
+        return "REAL"
+    # Otherwise, the affinity is NUMERIC. TODO: Precision and scale
+    return "NUMERIC"
+
+
+def sqlite_connection_to_introspection_result(con: sqlite3.Connection) -> IntrospectionResult:
+    schema: IntrospectionResult = cast(IntrospectionResult, {})
+    for (
+        table_name,
+        column_id,
+        column_name,
+        column_type,
+        _notnull,
+        _default_value,
+        pk,
+    ) in query_connection(con, LIST_TABLES_QUERY, {}):
+        table = schema.get(table_name, ([], TableParams({})))
+        cast(Tuple[List[TableColumn], TableParams], table)[0].append(
+            TableColumn(column_id + 1, column_name, sqlite_to_postgres_type(column_type), pk != 0)
+        )
+        schema[table_name] = table
+    return schema
+
+
+class SQLiteDataSource(LoadableDataSource):
+
+    table_params_schema: Dict[str, Any] = {"type": "object", "properties": {}}
+
+    params_schema: Dict[str, Any] = {
+        "type": "object",
+        "properties": {
+            "url": {
+                "type": "string",
+                "description": "HTTP URL to the sqlite file",
+                "title": "URL",
+            }
+        },
+    }
+
+    supports_mount = False
+    supports_load = True
+    supports_sync = False
+
+    _icon_file = "sqlite.svg"  # TODO
+
+    def _load(self, schema: str, tables: Optional[TableInfo] = None):
+        with db_from_minio(str(self.params.get("url"))) as con:
+            introspection_result = sqlite_connection_to_introspection_result(con)
+            for table_name, table_definition in introspection_result.items():
+                schema_spec = cast(Tuple[List[TableColumn], TableParams], table_definition)[0]
+                self.engine.create_table(
+                    schema=schema,
+                    table=table_name,
+                    schema_spec=schema_spec,
+                )
+                table_contents = query_connection(
+                    con, 'SELECT * FROM "{}"'.format(sql_quote_str(table_name))  #  nosec
+                )
+                self.engine.run_sql_batch(
+                    SQL("INSERT INTO {0}.{1} ").format(Identifier(schema), Identifier(table_name))
+                    + SQL(" VALUES (" + ",".join(itertools.repeat("%s", len(schema_spec))) + ")"),
+                    # TODO: break this up into multiple batches for larger sqlite files
```

Comment:
Probably a pretty high-priority optimization post-MVP, more so than downloading the entire file on introspection (downloading a 100MB file will be a few seconds at most but loading the whole table in-memory as Python tuples would incur a pretty high RAM cost)

Reply:
Agreed, this is probably the first thing that will break when someone tries to upload a decent-sized sqlite file.